zetaclient/chains/solana/signer/execute.go

```diff
@@ -43,12 +46,91 @@ func (signer *Signer) prepareExecuteTx(
 			return nil, errors.Wrap(err, "error creating execute instruction")
 		}
 
-		return signer.createOutboundWithFallback(ctx, inst, msgIn, params.CallOptions.GasLimit)
+		return signer.createOutboundWithFallback(
+			ctx,
+			inst,
+			msgIn,
+			params.CallOptions.GasLimit,
+			msg.ALT(),
+			msg.ALTStateAddresses(),
+		)
 	}, nil
 }
 
+func (signer *Signer) prepareExecuteMsg(
+	cctx *types.CrossChainTx,
+) (contracts.ExecuteType, *contracts.GenericExecuteMsg, error) {
+	var executeType contracts.ExecuteType
+	if cctx.CctxStatus.Status == types.CctxStatus_PendingRevert && cctx.RevertOptions.CallOnRevert {
+		executeType = contracts.ExecuteTypeRevert
+	} else {
+		executeType = contracts.ExecuteTypeCall
+	}
+
+	var message []byte
+	if executeType == contracts.ExecuteTypeRevert {
+		message = cctx.RevertOptions.RevertMessage
+	} else {
+		messageToDecode, err := hex.DecodeString(cctx.RelayedMessage)
+		if err != nil {
+			return executeType, nil, errors.Wrapf(err, "decodeString %s error", cctx.RelayedMessage)
+		}
+		message = messageToDecode
+	}
+
+	msg, err := contracts.DecodeExecuteMsg(message)
+	if err != nil {
+		return executeType, nil, errors.Wrapf(err, "decode ExecuteMsg %s error", cctx.RelayedMessage)
+	}
+
+	return executeType, msg, nil
+}
+
+func (signer *Signer) prepareExecuteMsgParams(
+	ctx context.Context,
+	msg *contracts.GenericExecuteMsg,
+) ([]*solana.AccountMeta, solana.PublicKeySlice, error) {
+	remainingAccounts := []*solana.AccountMeta{}
+	if msg.ALTAddress() == nil {
+		for _, a := range msg.Legacy.Accounts {
+			remainingAccounts = append(remainingAccounts, &solana.AccountMeta{
+				PublicKey:  solana.PublicKey(a.PublicKey),
+				IsWritable: a.IsWritable,
+			})
+		}
+
+		return remainingAccounts, nil, nil
+	}
+
+	alt, err := addresslookuptable.GetAddressLookupTableStateWithOpts(
+		ctx,
+		signer.client.(*rpc.Client),
+		*msg.ALTAddress(),
+		&rpc.GetAccountInfoOpts{Commitment: rpc.CommitmentProcessed},
+	)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "cannot get alt")
+	}
+
+	writableSet := make(map[int]struct{}, len(msg.Alt.WriteableIndexes))
+	for _, j := range msg.Alt.WriteableIndexes {
+		writableSet[int(j)] = struct{}{}
```

Comment:
Do we need to validate the value of `j` is in the range of `[0, len(alt.Addresses))`?  Otherwise it's a invalid tx?

Reply:
it is up to user, it will fail and be replaced with fallback increment nonce

otherwise if we stop it here this will get retried since it is in prepare part, not sure we have cancellation in this part for solana yet, just increment nonce fallback need to check